Initial support for half precision

e10harvey opened the pull request:
Initial support for half precision on intel/amd hosts and CUDA. Half precision toolchain support is printed at configure time but will fall back to single precision if a half precision intrinsic is not found.

    cmake: Various updates for half precision support
      - Added checks for FP16 support in host toolchains
        and cuda toolchains
      - Added HAVE_KOKKOSKERNELS_HALFMATH, HAVE_KOKKOSKERNELS_FP16,
        and HAVE_KOKKOSKERNELS_CUDA_FP16 defines
    
    src: Various updates for half precision support
      - Added Kokkos_HalfPrecision to typedef 'half' based on cmake checks
        - Added 'device_fp16_t' and 'host_fp16_t' types.
      - Updated Kokkos_ArithTraits to support 'half'
    
    unit_test/batched:
      - Add half precision tests for TeamGemm, TeamVectorGemm, and SerialGemm
      - Fixes #813
    
    test_common:
      - Add half precision ArithTraits tests
      - Fixes #814
      - Add vanilla gemm implementation for unit tests

e10harvey commented on src/batched/KokkosBatched_Gemm_TeamVector_Impl.hpp:
```diff
@@ -138,7 +138,7 @@ namespace KokkosBatched {
       // C = beta C + alpha A B
       // C (m x n), A(m x k), B(k x n)
-      return TeamGemmInternal<Algo::Gemm::Unblocked>::
+      return TeamVectorGemmInternal<Algo::Gemm::Unblocked>::
```
@kyungjoo-kim: Was this just a typo or is there a reason to be calling `TeamGemm` instead of `TeamVectorGemm` here?

kyungjoo-kim replied:
typo.

e10harvey commented:
@ndellingwood, are these half precision runtimes going to cause problems for our regression tests?
```bash
Note: Google Test filter = *half*:*gemm*float*
[==========] Running 24 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 24 tests from serial
[ RUN      ] serial.batched_scalar_serial_gemm_nt_nt_half_half
[       OK ] serial.batched_scalar_serial_gemm_nt_nt_half_half (2342 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_t_nt_half_half
[       OK ] serial.batched_scalar_serial_gemm_t_nt_half_half (2373 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_nt_t_half_half
[       OK ] serial.batched_scalar_serial_gemm_nt_t_half_half (2412 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_t_t_half_half
[       OK ] serial.batched_scalar_serial_gemm_t_t_half_half (2436 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_nt_nt_float_float
[       OK ] serial.batched_scalar_serial_gemm_nt_nt_float_float (330 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_t_nt_float_float
[       OK ] serial.batched_scalar_serial_gemm_t_nt_float_float (335 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_nt_t_float_float
[       OK ] serial.batched_scalar_serial_gemm_nt_t_float_float (324 ms)
[ RUN      ] serial.batched_scalar_serial_gemm_t_t_float_float
[       OK ] serial.batched_scalar_serial_gemm_t_t_float_float (332 ms)
<snip>
```

ndellingwood commented:
> are these half precision runtimes going to cause problems for our regression tests?

@e10harvey no, those times should be fine. If we run into a situation with timeouts due to overlapping tests on the same resource we'll adjust the nightly test launch times to make sure we don't self-inflict issues.

e10harvey commented on unit_test/batched/Test_Batched_TeamGemm_Real.hpp:
```diff
@@ -1,47 +1,73 @@
+#if defined(HAVE_KOKKOSKERNELS_HALFMATH)
```
Note: This may run with `::Test::halfScalarType = float` if half precision is available on only the host or device.

crtrott replied:
whats the difference between test_batched_teamgemm and test_batched_teamgemm_half?

e10harvey replied:
the main difference is that test_batched_teamgemm_half always does the following in single precision:
1. generates random numbers
2. checks results on host

since the half precision intrinsic may not be compatible on both host and device

e10harvey commented on unit_test/batched/Test_Batched_TeamVectorGemm_Real.hpp:
```diff
@@ -0,0 +1,80 @@
+#if defined(HAVE_KOKKOSKERNELS_HALFMATH)
+TEST_F( TestCategory, batched_scalar_team_vector_gemm_nt_nt_half_half ) {
```
Note: This may run with `::Test::halfScalarType = float` if half precision is available on only the host or device.

e10harvey commented on unit_test/batched/Test_Batched_SerialGemm_Real.hpp:
```diff
@@ -1,3 +1,30 @@
+#if defined(HAVE_KOKKOSKERNELS_HALFMATH)
```
Note: This may run with `::Test::halfScalarType = float` if half precision is available on only the host or device.

e10harvey commented:
@srajama1: This PR is ready for review.

## Summary of recent changes:
- Replaced `KOKKOSKERNEL_INST_FLOAT` with `HAVE_KOKKOSKERNELS_HALFMATH`. `HAVE_KOKKOSKERNELS_HALFMATH` is set by `cmake/HalfPrecisionSupport.cmake` if the toolchain supports half precision.
- Reverted `UTV2` changes and added `TeamVectorGemm` unit tests in `unit_test/batched/Test_Batched_TeamVectorGemm.hpp`.
- Updated all half precision batched gemm unit tests to:
  - test both the blocked and unblocked algorithms
  - test batched algorithms against a vanilla gemm reference implementation in `test_common/KokkosKernels_TestUtils.hpp`
- Added half precision tests in `test_commn/Test_Common_ArithTraits.hpp` for `src/Kokkos_ArithTraits.hpp`
- Found and fixed #813 and #814

## Spot-checks:
```
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=674 run_time=102
clang-8.0-Pthread_Serial-release build_time=218 run_time=119
clang-9.0.0-Pthread-release build_time=121 run_time=56
clang-9.0.0-Serial-release build_time=135 run_time=47
cuda-10.1-Cuda_OpenMP-release build_time=914 run_time=143
gcc-7.3.0-OpenMP-release build_time=147 run_time=45
gcc-7.3.0-Pthread-release build_time=116 run_time=56
gcc-8.3.0-Serial-release build_time=145 run_time=50
gcc-9.1-OpenMP-release build_time=181 run_time=46
gcc-9.1-Serial-release build_time=167 run_time=49
intel-18.0.5-OpenMP-release build_time=364 run_time=44
intel-19.0.5-Pthread-release build_time=414 run_time=55
#######################################################
FAILED TESTS
#######################################################
intel-17.0.1-Serial-release (build failed)
#######################################################
```

Note that half precision was only tested on cuda-10.1 here:
```bash
$ for file in $(find -name '*.configure.log'); do grep -H 'HAVE.*FP16' $file; done
./TestAll_2020-09-23_12.14.15/cuda/10.1/Cuda_OpenMP-release/cuda-10.1-Cuda_OpenMP-release.configure.log:   HAVE_FP16:      FALSE
./TestAll_2020-09-23_12.14.15/cuda/10.1/Cuda_OpenMP-release/cuda-10.1-Cuda_OpenMP-release.configure.log:   HAVE_CUDA_FP16: TRUE
./TestAll_2020-09-23_13.29.24/cuda/10.1/Cuda_OpenMP-release/cuda-10.1-Cuda_OpenMP-release.configure.log:   HAVE_FP16:      FALSE
./TestAll_2020-09-23_13.29.24/cuda/10.1/Cuda_OpenMP-release/cuda-10.1-Cuda_OpenMP-release.configure.log:   HAVE_CUDA_FP16: TRUE
```

For intel/17, the first error is:
```
/path/to/kokkos/core/src/impl/Kokkos_EBO.hpp(241): error: cannot overload functions distinguished by return type alone
    constexpr T const& no_unique_address_data_member() const& {
```

```
#######################################################
PASSED TESTS
#######################################################
arm-20.0-OpenMP-release build_time=379 run_time=93
arm-20.0-OpenMP_Serial-release build_time=642 run_time=189
arm-20.0-Serial-release build_time=355 run_time=87
```

Note that half precision was tested on both host and device spaces here:
```
./TestAll_2020-09-23_13.43.36/arm/20.0/OpenMP-release/arm-20.0-OpenMP-release.configure.log:   HAVE_FP16:      TRUE
./TestAll_2020-09-23_13.43.36/arm/20.0/OpenMP-release/arm-20.0-OpenMP-release.configure.log:   HAVE_CUDA_FP16: FALSE
./TestAll_2020-09-23_13.43.36/arm/20.0/Serial-release/arm-20.0-Serial-release.configure.log:   HAVE_FP16:      TRUE
./TestAll_2020-09-23_13.43.36/arm/20.0/Serial-release/arm-20.0-Serial-release.configure.log:   HAVE_CUDA_FP16: FALSE
./TestAll_2020-09-23_13.43.36/arm/20.0/OpenMP_Serial-release/arm-20.0-OpenMP_Serial-release.configure.log:   HAVE_FP16:      TRUE
./TestAll_2020-09-23_13.43.36/arm/20.0/OpenMP_Serial-release/arm-20.0-OpenMP_Serial-release.configure.log:   HAVE_CUDA_FP16: FALSE
```

rsiva-web commented:
@crtrott : I would like your view on this PR from Kokkos, C++ standards and compilers perspective.

lucbv commented:
@srajama1 @e10harvey 
This will not work on HIP backend, do we want to fix it there first or wait for this to merge and then fix it?

e10harvey commented:
> This will not work on HIP backend, do we want to fix it there first or wait for this to merge and then fix it?

@lucbv: I have only targeted intel/amd hosts and CUDA; I'll defer to @srajama1.

lucbv commented:
@e10harvey completely agree I just wanted to mention it : )

e10harvey commented:
> @e10harvey completely agree I just wanted to mention it : )

@lucbv Thanks for mentioning it :) I've updated the description to capture this.

srajama1 commented:
Can we check for AMD GPU ARCH and disable half precision support for those ?

e10harvey commented:
> Can we check for AMD GPU ARCH and disable half precision support for those ?

Certainly. It will currently be disabled as `HAVE_KOKKOSKERNELS_HALFMATH` will not be defined and `half` will be an alias for `float`. This will allow us to start writing and testing kernels for AMD GPU ARCH and later wire in support for half precision. If this should be handled differently, please let me know.

crtrott commented on src/Kokkos_HalfPrecision.hpp:
```diff
@@ -0,0 +1,173 @@
+            return ret;
+        }
+#endif
+        // float
```
The following overloads feel funky. Even if they are temporary workaround you may get into trouble. Consider just having generic functions:
```
half cast_to_half(double); half cast_to_half(float); half cast_to_half(half);
```

crtrott replied:
With trouble I mean later, when you try to get rid of them again.

crtrott replied:
Also generally having a cast function both take a return argument and return it is pretty iffy from a design perspective.

crtrott replied:
if you need the other direction add cast_to_double etc. Note I get that this float2half business is how CUDA names the cast functions but there are not different overloads of those functions. I would REALLY recommend to change this in particular since you make them public. We could also deal with this inside of Kokkos Core and do a proper solution for you.

e10harvey replied:
agreed. these overloads can actually be removed since GEMM is the only kernel being used with `half`. the overloads were added [here](https://github.com/e10harvey/kokkos-kernels/blob/6eda84d2b9cd675cc182385d6627297ceb53a961/src/batched/KokkosBatched_FindAmax_Internal.hpp#L62) as preliminary work around for [this](https://github.com/e10harvey/kokkos-kernels/blob/6eda84d2b9cd675cc182385d6627297ceb53a961/src/batched/KokkosBatched_FindAmax_Internal.hpp#L50-L53). i'd be interested in helping with the proper solution within Kokkos Core. I will remove these overloads unless I hear otherwise.

crtrott commented on test_common/KokkosKernels_TestUtils.hpp:
```diff
@@ -105,6 +105,94 @@ namespace Test {
+  #endif // KOKKOS_ENABLE_CUDA
+
+  template<class ViewTypeA, class ViewTypeB, class ViewTypeC, class ExecutionSpace>
+  struct SharedVanillaGEMM {
```
Why is this Functor in TestUtils?

e10harvey replied:
in an effort to reduce maintenance by providing a single reference GEMM implementation for other test authors to use. should this be moved outside of TestUtils?

crtrott commented on test_common/Test_Common_ArithTraits.hpp:
```diff
@@ -910,37 +920,37 @@ class ArithTraitsTesterTranscendentalBase<ScalarType, DeviceType, 1> :
     if (!equal(AT::atan (AT::tan (three)), three)) {
       printf ("AT::atan(tan(3)) != 3\n");
-      success = 0;
+      FAILURE();
```
drive-by changes?

crtrott commented on test_common/Test_Common_ArithTraits.hpp:
```diff
@@ -1627,8 +1679,8 @@ void test_ArithTraits ()
   NullBuffer null_buffer;
-  //std::ostream &out = std::cout;
-  std::ostream out(&null_buffer);
+  std::ostream &out = std::cerr;
```
intended change?

e10harvey replied:
yes

crtrott commented on unit_test/batched/Test_Batched_SerialGemm.hpp:
```diff
@@ -66,6 +66,96 @@ namespace Test {
+        typename ScalarType,
+        typename ParamTagType, 
+        typename AlgoTagType>
+  void impl_test_batched_gemm_half(const int N, const int matAdim1, const int matAdim2, const int matBdim1, const int matBdim2,
```
why is this called impl_test_batched_half when nothing in this function actually refers to whether anything is half precision?

crtrott replied:
actually it looks like there is FP16_EPSILON? Why not just make a conditional there if the rest of the code is the same as the "non-half" version?

e10harvey replied:
> why is this called impl_test_batched_half when nothing in this function actually refers to whether anything is half precision?

same as https://github.com/kokkos/kokkos-kernels/pull/794#discussion_r496779387.



> actually it looks like there is FP16_EPSILON? Why not just make a conditional there if the rest of the code is the same as the "non-half" version?

we could use the same approach with respect to always generating random numbers in single precision and checking results on the host in single precision. would this be an acceptable approach for complex numbers?

crtrott commented on unit_test/batched/Test_Batched_SerialGemm.hpp:
```diff
@@ -187,3 +277,65 @@ int test_batched_gemm() {
+         typename ScalarType,
+         typename ParamTagType,
+         typename AlgoTagType>
+int test_batched_gemm_half() {
```
again why the name if nothing actually requires this to be half precision? I mean you could instantiate this with double precision views or?

e10harvey replied:
same as https://github.com/kokkos/kokkos-kernels/pull/794#discussion_r496779387. you could instantiate this with any scalar type.

srajama1 replied:
Then the name shouldn't have half, right ?

e10harvey replied:
I will remove this function and add branches via `std::is_same<ScalarType,half_t` to `impl_test_batched_gemm_half`. @srajama1: Is there anything else that you'd like addressed?

crtrott commented on unit_test/batched/Test_Batched_TeamGemm.hpp:
```diff
@@ -130,67 +130,220 @@ namespace Test {
+           typename ScalarType,
+           typename ParamTagType, 
+           typename AlgoTagType>
+  void impl_test_batched_teamgemm_half(const int N, const int matAdim1, const int matAdim2, const int matBdim1, const int matBdim2,
```
same thing again why the name if this is a generic function?

e10harvey replied:
same as above :)

crtrott commented on src/Kokkos_HalfPrecision.hpp:
```diff
@@ -0,0 +1,173 @@
+#if defined(HAVE_KOKKOSKERNELS_FP16)
+        static inline
+        float half2float(host_fp16_t x, float &ret) {
+            ret = __cast2float(x);
```
Where is this overload supposed to come from if HAVE_KOKKOSKERNELS_CUDA_FP16 is true?

e10harvey replied:
the overload is defined only here if HAVE_KOKKOSKERNELS_FP16 is true. so if both HAVE_KOKKOSKERNELS_FP16 and HAVE_KOKKOSKERNELS_CUDA_FP16 are true, the overload will come from here and `half` will default to device_fp16_t.

crtrott reviewed (changes requested):
I put down a number of comments. The biggest issue is with the casting functions, I think those should change. Furthermore I don't think the logic right now works with both Host and CUDA FP16 enabled (overloads seem to be missing). 

Another issue is the introduction of types and typedefs which kinda only work on one side. Note this has the clear potential to screw up functors and lambdas.  I think you can't have scalar half members either right now as members of functors or capture scalar half in a lambda. Potentially this will either screw up the sizes of the classes (they are different size on host and device which means the kernel launch will corrupt data) or you have unusable types on one side. 

There might be more problems but this was all I clearly identified within like 45 mins review.

crtrott commented:
Generally the `__half` type seems to exist on both host and device, so do all the conversion functions. Only the math functions don't but we should be able to protect those like this, if we don't have a native host half precision support.
```c++
__host__ __device__ __half abs(__half val) {
#ifdef __CUDA_ARCH__
return __habs(val);
#else
return __float2half(fabs(__half2float(val)));
#endif
}
```

If we do and the bit format is the same we could also work with e.g.
```c++
__host__ __device__ __half abs(__half val) {
#ifdef __CUDA_ARCH__
return __habs(val);
#else
return reinterprete_cast<__half>(__mkl_habs(reinterprete_cast<__mkl_half&>(val)));
#endif
}
```

This would remove all the mess about distinguishing between host and device types from the user perspective.

crtrott commented:
I started working on putting half precision support into Kokkos see here: https://github.com/kokkos/kokkos/pull/3439

This just adds the conversions.

srajama1 commented:
@e10harvey Should we update this PR with Kokkos changes ? Jennifer Loe would like to use it.

e10harvey commented:
> @e10harvey Should we update this PR with Kokkos changes ? Jennifer Loe would like to use it.

Once they are merged, I will update this PR. But, the Kokkos changes are still in progress.

srajama1 commented:
@crtrott : Any update on Kokkos merging ?

e10harvey commented:
@srajama1 and @crtrott: Thank you for your feedback! This PR, https://github.com/crtrott/kokkos/pull/10, and https://github.com/kokkos/kokkos/pull/3439 are ready to review.

## Spot-check
```
KokkosKernels Repository Status:  f0eafbd96bdcc94c1999a37ee63f98c43b7bfd9a Ensure half_t is device type in ArithTraits

Kokkos Repository Status:  43bb63d98f0a408f8ad573ec5ff3e3ba14c1764a CI workarounds and fixes
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=663 run_time=134
clang-8.0-Pthread_Serial-release build_time=207 run_time=131
clang-9.0.0-Pthread-release build_time=121 run_time=62
clang-9.0.0-Serial-release build_time=130 run_time=49
cuda-10.1-Cuda_OpenMP-release build_time=858 run_time=136
cuda-11.0-Cuda_OpenMP-release build_time=911 run_time=136
cuda-9.2-Cuda_Serial-release build_time=829 run_time=207
gcc-7.3.0-OpenMP-release build_time=152 run_time=47
gcc-7.3.0-Pthread-release build_time=122 run_time=62
gcc-8.3.0-Serial-release build_time=145 run_time=55
gcc-9.1-OpenMP-release build_time=188 run_time=48
gcc-9.1-Serial-release build_time=166 run_time=52
intel-17.0.1-Serial-release build_time=377 run_time=69
intel-18.0.5-OpenMP-release build_time=434 run_time=58
intel-19.0.5-Pthread-release build_time=452 run_time=74
```

## Spot-check-tpls
```
KokkosKernels Repository Status:  f0eafbd96bdcc94c1999a37ee63f98c43b7bfd9a Ensure half_t is device type in ArithTraits

Kokkos Repository Status:  43bb63d98f0a408f8ad573ec5ff3e3ba14c1764a CI workarounds and fixes
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=641 run_time=136
clang-8.0-Pthread_Serial-release build_time=204 run_time=135
clang-9.0.0-Pthread-release build_time=131 run_time=62
clang-9.0.0-Serial-release build_time=131 run_time=50
cuda-10.1-Cuda_OpenMP-release build_time=912 run_time=141
cuda-11.0-Cuda_OpenMP-release build_time=974 run_time=141
gcc-7.3.0-OpenMP-release build_time=149 run_time=49
gcc-7.3.0-Pthread-release build_time=116 run_time=63
gcc-8.3.0-Serial-release build_time=148 run_time=54
gcc-9.1-OpenMP-release build_time=182 run_time=48
gcc-9.1-Serial-release build_time=164 run_time=52
intel-17.0.1-Serial-release build_time=288 run_time=53
intel-18.0.5-OpenMP-release build_time=369 run_time=48
intel-19.0.5-Pthread-release build_time=410 run_time=63
```

crtrott reviewed (changes requested):
I still think the test names are weird but its fine. The macros are not fine however. In general you need to prefix all macros you introduce in particular in the library since otherwise you may get conflicts.

e10harvey commented:
@crtrott: Thank you for your reviews. Is there anything else you'd like addressed?

crtrott commented:
Otherwise it looks ok for me. Note: I only looked for stuff which clearly stands out to me as problematic, I didn't look whether this is in line with general KokkosKernels coding standards etc.

e10harvey commented:
@ndellingwood, @srajama1: We are still iterating on https://github.com/kokkos/kokkos/pull/3439 so this PR may also require some changes. Otherwise, is there anything else you'd like addressed?

e10harvey commented:
@ndellingwood, @srajama1: This is ready for a final review. Is there anything else you'd like addressed?

e10harvey commented:
Latest spot-check results:

## Spot-check
```
KokkosKernels Repository Status:  19e1eb1e93bb94adde5704bd9b0b79fcc936ad5a Remove HALF_IMPL_TYPE

Kokkos Repository Status:  9a875b43938ceacd6ced5cdb9abef96d84d78142 Define macro if Cuda has half_t
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=641 run_time=134
clang-8.0-Pthread_Serial-release build_time=205 run_time=132
clang-9.0.0-Pthread-release build_time=124 run_time=61
clang-9.0.0-Serial-release build_time=132 run_time=50
cuda-10.1-Cuda_OpenMP-release build_time=861 run_time=136
cuda-11.0-Cuda_OpenMP-release build_time=918 run_time=136
cuda-9.2-Cuda_Serial-release build_time=831 run_time=207
gcc-7.3.0-OpenMP-release build_time=148 run_time=47
gcc-7.3.0-Pthread-release build_time=119 run_time=62
gcc-8.3.0-Serial-release build_time=145 run_time=52
gcc-9.1-OpenMP-release build_time=184 run_time=46
gcc-9.1-Serial-release build_time=164 run_time=51
intel-17.0.1-Serial-release build_time=283 run_time=54
intel-18.0.5-OpenMP-release build_time=385 run_time=49
intel-19.0.5-Pthread-release build_time=394 run_time=63
```

## Spot-check-tpls
```
KokkosKernels Repository Status:  19e1eb1e93bb94adde5704bd9b0b79fcc936ad5a Remove HALF_IMPL_TYPE

Kokkos Repository Status:  9a875b43938ceacd6ced5cdb9abef96d84d78142 Define macro if Cuda has half_t
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=639 run_time=137
clang-8.0-Pthread_Serial-release build_time=203 run_time=135
clang-9.0.0-Pthread-release build_time=132 run_time=62
clang-9.0.0-Serial-release build_time=130 run_time=50
cuda-10.1-Cuda_OpenMP-release build_time=933 run_time=142
cuda-11.0-Cuda_OpenMP-release build_time=1051 run_time=335
gcc-7.3.0-OpenMP-release build_time=149 run_time=48
gcc-7.3.0-Pthread-release build_time=124 run_time=64
gcc-8.3.0-Serial-release build_time=145 run_time=53
gcc-9.1-OpenMP-release build_time=185 run_time=49
gcc-9.1-Serial-release build_time=161 run_time=53
intel-17.0.1-Serial-release build_time=288 run_time=54
intel-18.0.5-OpenMP-release build_time=387 run_time=48
intel-19.0.5-Pthread-release build_time=399 run_time=63
```

rsiva-web commented:
It would be nice to get this before the release. @ndellingwood : Can you also check these changes once ?

ndellingwood commented:
@srajama1 sounds good, I'll take a scan later today

srajama1 reviewed (commented):
Two comments below.

e10harvey commented on cmake/HalfPrecisionSupport.cmake:
```diff
@@ -0,0 +1,6 @@
+# Check whether Kokkos has half precision headers
+IF(EXISTS ${Kokkos_DIR}/../../../include/Kokkos_Half.hpp)
```
home. this just checks for the existence of Kokkos_Half.hpp in Kokkos.

e10harvey replied:
@jjwilke: can you weigh in on approaches for setting `HAVE_KOKKOS_HALFMATH`?

ndellingwood commented:
@e10harvey I don't have much to add to the source code review, I have a couple questions/comments:

- `HAVE_KOKKOS_HALFMATH` will be set automatically if Kokkos' half precision support is available, detected by checking the directory pointed at in comments above is available https://github.com/kokkos/kokkos-kernels/pull/794#discussion_r516853549 ; just to confirm, the plan is to have half-precision always available when it is available through Kokkos, rather than a configuration option?

- There may be some flux in what is supported/tested for half-precision support (e.g. kokkos/kokkos#3549, kokkos/kokkos#3561), should we wait to merge this until some dust settles following merge of kokkos/kokkos#3439 (and we update our testing accordingly)?

e10harvey commented:
>     * `HAVE_KOKKOS_HALFMATH` will be set automatically if Kokkos' half precision support is available, detected by checking the directory pointed at in comments above is available [#794 (comment)](https://github.com/kokkos/kokkos-kernels/pull/794#discussion_r516853549) ; just to confirm, the plan is to have half-precision always available when it is available through Kokkos, rather than a configuration option?

Correct.

>     * There may be some flux in what is supported/tested for half-precision support (e.g. [kokkos/kokkos#3549](https://github.com/kokkos/kokkos/issues/3549), [kokkos/kokkos#3561](https://github.com/kokkos/kokkos/issues/3561)), should we wait to merge this until some dust settles following merge of [kokkos/kokkos#3439](https://github.com/kokkos/kokkos/pull/3439) (and we update our testing accordingly)?

Yes.

ndellingwood commented:
Cross-referencing PR kokkos/kokkos#3563

e10harvey commented:
Latest spot-check results:
## Spot-check
```
KokkosKernels Repository Status:  45d2237e9b62e6a7216cfe47399499784d5c2e0d Replace HAVE_KOKKOS_HALFMATH with KOKKOS_HALF_T_IS_FLOAT

Kokkos Repository Status:  b60c37fa6028830f602ed8b1223ec3d7b659650e Fix half_t for KEPLER35
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=654 run_time=134
clang-8.0-Pthread_Serial-release build_time=216 run_time=132
clang-9.0.0-Pthread-release build_time=130 run_time=62
clang-9.0.0-Serial-release build_time=135 run_time=50
cuda-10.1-Cuda_OpenMP-release build_time=889 run_time=138
cuda-11.0-Cuda_OpenMP-release build_time=943 run_time=147
cuda-9.2-Cuda_Serial-release build_time=887 run_time=218
gcc-7.3.0-OpenMP-release build_time=151 run_time=47
gcc-7.3.0-Pthread-release build_time=127 run_time=62
gcc-8.3.0-Serial-release build_time=148 run_time=53
gcc-9.1-OpenMP-release build_time=185 run_time=47
gcc-9.1-Serial-release build_time=168 run_time=51
intel-17.0.1-Serial-release build_time=295 run_time=58
intel-18.0.5-OpenMP-release build_time=488 run_time=47
intel-19.0.5-Pthread-release build_time=398 run_time=63
```

## Spot-check-tpls
```
KokkosKernels Repository Status:  45d2237e9b62e6a7216cfe47399499784d5c2e0d Replace HAVE_KOKKOS_HALFMATH with KOKKOS_HALF_T_IS_FLOAT

Kokkos Repository Status:  b60c37fa6028830f602ed8b1223ec3d7b659650e Fix half_t for KEPLER35
<snip>
#######################################################
PASSED TESTS
#######################################################
clang-8.0-Cuda_OpenMP-release build_time=717 run_time=140
clang-8.0-Pthread_Serial-release build_time=240 run_time=137
clang-9.0.0-Pthread-release build_time=153 run_time=63
clang-9.0.0-Serial-release build_time=162 run_time=51
cuda-10.1-Cuda_OpenMP-release build_time=1888 run_time=146
cuda-11.0-Cuda_OpenMP-release build_time=1810 run_time=118
gcc-7.3.0-OpenMP-release build_time=177 run_time=54
gcc-7.3.0-Pthread-release build_time=138 run_time=65
gcc-8.3.0-Serial-release build_time=165 run_time=54
gcc-9.1-OpenMP-release build_time=215 run_time=50
gcc-9.1-Serial-release build_time=193 run_time=54
intel-17.0.1-Serial-release build_time=396 run_time=54
intel-18.0.5-OpenMP-release build_time=699 run_time=49
intel-19.0.5-Pthread-release build_time=354 run_time=63
```

srajama1 reviewed (approved):
Looks good to me ! Thanks @e10harvey !

e10harvey commented:
@crtrott: Merging is blocked due to un-addressed requested changes. I addressed your feedback via https://github.com/kokkos/kokkos-kernels/pull/794/commits/57a0c83d78cbbdbf571bacef2040bcbdfe178d56. Is there anything else that you'd like addressed?

crtrott commented on src/KokkosKernels_Half.hpp:
```diff
@@ -0,0 +1,67 @@
+#ifndef KOKKOSKERNELS_HALF_HPP
+#define KOKKOSKERNELS_HALF_HPP
+
+#if defined(KOKKOS_HALF_T_IS_FLOAT)
```
sorry had an outstanding review, but we talked on slack about htis. So its fine.

e10harvey replied:
thanks for catching this!
